File: python/file_size.py
#! /usr/bin/env python
"""
This file is part of file_size.
Copyright (C) 2024  xkk1

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>.
"""
import datetime
import os
import pathlib
import sys


def get_file_infomation(file_path: str) -> dict:
    """
    获取文件信息
    """
    file_infomation: dict = {
        "code": 0,
        "message": "success",
        "input_path": file_path,
        "type": "unknow",
        "size": 0
    }
    try:
        path = pathlib.Path(file_path)
        # 文件名/目录名
        file_infomation["name"] = path.name
        # 路径
        file_infomation["path"] = str(path)
        # 绝对路径
        file_infomation["absolute_path"] = str(path.absolute())
        # 获取文件信息
        stat_result: os.stat_result = path.stat()
        # 文件大小
        file_infomation["size"] = stat_result.st_size
        # 访问时间
        file_infomation["access_time"] = stat_result.st_atime
        # 修改时间
        file_infomation["modify_time"] = stat_result.st_mtime
        # 创建时间
        try:
            # https://docs.python.org/zh-cn/3/library/os.html#os.stat_result.st_birthtime
            # 该属性并不总是可用的，并可能引发 AttributeError
            # 在 3.12 版本发生变更: 目前 st_birthtime 已在 Windows 上可用
            file_infomation["create_time"] = stat_result.st_birthtime
        except:
            file_infomation["create_time"] = None
    except Exception as e:
        file_infomation["code"] = 1
        file_infomation["msg"] = str(e)
        return file_infomation
    
    if not path.exists():
        file_infomation["code"] = 2
        file_infomation["msg"] = "文件不存在"
        return file_infomation
    
    if path.is_file():
        file_infomation["type"] = "file"  # 文件
        # 扩展名
        file_infomation["extension"] = path.suffix[1:] if path.suffix else ""
        return file_infomation

    if path.is_dir():
        file_infomation["type"] = "dir"  # 目录
        file_infomation["iter_path"] = []
        file_infomation["size"] = 0
        for iter_path in path.iterdir():
            iter_path_infomation: dict = get_file_infomation(iter_path)
            file_infomation["iter_path"].append(iter_path_infomation)
            if iter_path_infomation["code"] == 0:
                file_infomation["size"] += iter_path_infomation["size"]
        return file_infomation

    file_infomation["code"] = -1
    file_infomation["msg"] = "不支持的文件类型"
    return file_infomation

def get_file_size_info(file_size_bytes):
    file_size_info: str = ""

    # IEC KiB    1024 B = 1 KiB
    # 二进制词头 https://zh.wikipedia.org/wiki/%E4%BA%8C%E9%80%B2%E4%BD%8D%E5%89%8D%E7%BD%AE%E8%A9%9E
    # IEC 60027-2 https://zh.wikipedia.org/wiki/IEC_60027-2
    binary_prefix: str = "KMGTPEZYRQ"  # TiB, GiB, MiB, KiB
    # SI kB    1000 B = 1 kB
    # 国际单位制词头 https://zh.wikipedia.org/wiki/%E5%9B%BD%E9%99%85%E5%8D%95%E4%BD%8D%E5%88%B6%E8%AF%8D%E5%A4%B4
    # 十进制前缀 SI https://zh.wikipedia.org/wiki/%E5%9B%BD%E9%99%85%E5%8D%95%E4%BD%8D%E5%88%B6
    metric_prefix: str = "kMGTPEZYRQ"  # TB, GB, MB, kB
    metric_prefix_zh: str = "千兆吉太拍艾泽尧容昆"
    # JEDEC KB    1024 B = 1 KB
    # JEDEC https://en.wikipedia.org/wiki/JEDEC_memory_standards
    jedec_prefix: str = binary_prefix  # TB, GB, MB, KB
    
    # B、KiB、MiB、GiB、TiB
    if file_size_bytes < 1024:
        file_size_info += f"{file_size_bytes} B"
    else:
        for i in range(1, len(binary_prefix)):
            if file_size_bytes < 1024 ** (i + 1):
                file_size_info += f"{file_size_bytes / (1024 ** i):.4f} {binary_prefix[i-1]}iB"
                break
        else:
            file_size_info += f"{file_size_bytes / (1024 ** len(binary_prefix)):.4f}"\
            + f"{binary_prefix[len(binary_prefix)-1]}iB"
    
    # B、KB、MB、GB、TB
    if file_size_bytes < 1000:
        file_size_info += " 字节"
    else:
        file_size_info += "    "
        for i in range(1, len(metric_prefix)):
            if file_size_bytes < 1000 ** (i + 1):
                file_size_info += f"{file_size_bytes / (1000 ** i):.4f} {metric_prefix[i-1]}B {metric_prefix_zh[i-1]}字节"
                break
        else:
            file_size_info += f"{file_size_bytes / (1000 ** len(metric_prefix)):.4f}"\
            + f"{metric_prefix[len(metric_prefix)-1]}B {metric_prefix_zh[len(metric_prefix)-1]}字节"
    
    # B 字节
    if file_size_bytes >= 1000:
        file_size_info += f"\n    {file_size_bytes} B 字节"
    
    # KiB    KB 千字节
    if not (1000 <= file_size_bytes <= 1024 * 1024) and file_size_bytes != 0:
        file_size_info += f"\n    {file_size_bytes / (1024):.4f} KiB"
        file_size_info += f"    {file_size_bytes / (1000):.4f} KB 千字节"
    
    # MiB    MB 兆字节
    if not (1000 * 1000 <= file_size_bytes <= 1024 * 1024 * 1024) and file_size_bytes >= 1000:
        file_size_info += f"\n    {file_size_bytes / (1024 ** 2):.4f} MiB"
        file_size_info += f"    {file_size_bytes / (1000 ** 2):.4f} MB 兆字节"
    
    # GiB    GB 吉字节
    if not (1000 * 1000 * 1000 <= file_size_bytes <= 1024 * 1024 * 1024 * 1024) and file_size_bytes >= 1000 * 1000:
        file_size_info += f"\n    {file_size_bytes / (1024 ** 3):.4f} GiB"
        file_size_info += f"    {file_size_bytes / (1000 ** 3):.4f} GB 吉字节"

    return file_size_info

def timestamp2str(timestamp: float) -> str:
    return datetime.datetime.fromtimestamp(timestamp).strftime(r"%Y-%m-%d %H:%M:%S")\
          + ("." + str(timestamp).split('.')[1]) if len(str(timestamp).split('.')) == 2 else ""

def get_file_infomation_str(file_path: str, return_size: bool=False) -> str:
    file_infomation: dict = get_file_infomation(file_path)
    file_infomation_str: str = ""
    file_infomation_str += f"{file_infomation['input_path']}\n"
    
    if file_infomation["code"] != 0:
        # 出错
        file_infomation_str += f"{file_infomation['msg']}\n"
    elif file_infomation["type"] == "file":
        # 文件
        file_infomation_str += f"文件名: {file_infomation['name']}\n"
        file_infomation_str += f"文件大小: {get_file_size_info(file_infomation['size'])}\n"
        file_infomation_str += f"扩展名: {file_infomation['extension']}\n"
        if file_infomation['path'] != file_infomation['absolute_path']:
            file_infomation_str += f"路径: {file_infomation['path']}\n"
        file_infomation_str += f"绝对路径: {file_infomation['absolute_path']}\n"
        file_infomation_str += f"访问时间: {timestamp2str(file_infomation['access_time'])}\n"
        file_infomation_str += f"修改时间: {timestamp2str(file_infomation['modify_time'])}\n"
        if file_infomation["create_time"]:
            file_infomation_str += f"创建时间: {timestamp2str(file_infomation['create_time'])}\n"
    elif file_infomation["type"] == "dir":
        # 目录
        file_infomation_str += f"目录名: {file_infomation['name']}\n"
        file_infomation_str += f"目录大小: {get_file_size_info(file_infomation['size'])}\n"
        if file_infomation['path'] != file_infomation['absolute_path']:
            file_infomation_str += f"路径: {file_infomation['path']}\n"
        file_infomation_str += f"绝对路径: {file_infomation['absolute_path']}\n"
        file_infomation_str += f"子文件数: {len(file_infomation['iter_path'])}\n"
        for i in range(len(file_infomation['iter_path'])):
            file_infomation_str += f"\n\n"
            file_infomation_str += get_file_infomation_str(file_infomation['iter_path'][i]['path'])
    else:
        pass
    if return_size:
        return file_infomation_str, file_infomation["size"]
    return file_infomation_str

def get_show_information() -> str:
    show_information: str = ""
    if len(sys.argv) == 1:
        show_information += f"""用法：{sys.argv[0] if ' ' not in sys.argv[0] else '"' + sys.argv[0] + '"'} [文件]...
显示文件大小信息

""" + get_file_infomation_str(sys.argv[0])
    elif len(sys.argv) == 2:
        show_information += get_file_infomation_str(sys.argv[1])
    else:
        total_file_size: int = 0
        for i in range(1, len(sys.argv)):
            information, size = get_file_infomation_str(sys.argv[i], return_size=True)
            total_file_size += size
            show_information += f"\n{''.ljust(20, '=')}\n\n"
            show_information += information
        show_information = \
            f"共输入了 {len(sys.argv) - 1} 个文件\n"\
            + f"总大小：{get_file_size_info(total_file_size)}"\
            + show_information
    return show_information

def main():
    print(get_show_information())


if __name__ == "__main__":
    main()

File: js/file-size.js
/*
This file is part of file-size.
Copyright (C) 2024  xkk1

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/
// 文件大小
"use strict";
// IEC KiB    1024 B = 1 KiB
// 二进制词头 https://zh.wikipedia.org/wiki/%E4%BA%8C%E9%80%B2%E4%BD%8D%E5%89%8D%E7%BD%AE%E8%A9%9E
// IEC 60027-2 https://zh.wikipedia.org/wiki/IEC_60027-2
let binary_prefix = "KMGTPEZYRQ";  // TiB, GiB, MiB, KiB
// SI kB    1000 B = 1 kB
// 国际单位制词头 https://zh.wikipedia.org/wiki/%E5%9B%BD%E9%99%85%E5%8D%95%E4%BD%8D%E5%88%B6%E8%AF%8D%E5%A4%B4
// 十进制前缀 SI https://zh.wikipedia.org/wiki/%E5%9B%BD%E9%99%85%E5%8D%95%E4%BD%8D%E5%88%B6
let metric_prefix = "kMGTPEZYRQ";  // TB, GB, MB, kB
let metric_prefix_zh = "千兆吉太拍艾泽尧容昆";
// JEDEC KB    1024 B = 1 KB
// JEDEC https://en.wikipedia.org/wiki/JEDEC_memory_standards
let jedec_prefix = binary_prefix;  // TB, GB, MB, KB

// 自动词缀
let auto_prefix = ["IEC 自动", "SI 自动", "JEDEC 自动", "SI 中文自动"];
// 所有支持的词缀
let all_prefix = ["B", "b"];
["B", "b"].forEach(b => {
    // IEC KiB    1024 B = 1 KiB
    for (let i = 0; i < binary_prefix.length; i++) {
        all_prefix.push(`IEC ${binary_prefix[i]}i${b}`);
    }
    // SI kB    1000 B = 1 kB
    for (let i = 0; i < metric_prefix.length; i++) {
        all_prefix.push(`SI ${metric_prefix[i]}${b}`);
    }
    // JEDEC KB    1024 B = 1 KB
    for (let i = 0; i < jedec_prefix.length; i++) {
        all_prefix.push(`JEDEC ${jedec_prefix[i]}${b}`);
    }
});
// SI 千字节    1000 字节 = 1 千字节
for (let i = 0; i < metric_prefix_zh.length; i++) {
    all_prefix.push(`SI ${metric_prefix_zh[i]}字节`);
    all_prefix.push(`SI ${metric_prefix_zh[i]}比特`);
}

// 上传文件列表
let fileList = [];

/**
 * 文件大小词缀转换为字节
 * @param {string} prefix 词缀 "B" "kB" "KiB"
 * @param {number} size 词缀大小，默认为1
 * @returns {number} 字节大小，未知词缀返回 -1
 */
function prefixToByte(prefix, size=1) {
    // prefix 不是字符或 size 不是数字，返回 0
    if (typeof prefix !== "string" || typeof size !== "number") return -1;
    if (prefix === "B" || prefix === "字节") return size;
    if (prefix === "b" || prefix === "比特") return size / 8;
    // IEC KiB    1024 B = 1 KiB
    for (let i = 0; i< binary_prefix.length; i++) {
        if (prefix === `${binary_prefix[i]}iB` || prefix === `IEC ${binary_prefix[i]}iB`) {
            return Math.pow(1024, i+1) * size;
        } else if (prefix === `${binary_prefix[i]}ib` || prefix === `IEC ${binary_prefix[i]}ib`) {
            return Math.pow(1024, i+1) * size / 8;
        }
    }
    // SI kB    1000 B = 1 kB
    for (let i = 0; i< metric_prefix.length; i++) {
        if (prefix === `${metric_prefix[i]}B` || prefix === `SI ${metric_prefix[i]}B`) {
            return Math.pow(1000, i+1) * size;
        } else if (prefix === `${metric_prefix[i]}b` || prefix === `SI ${metric_prefix[i]}b`) {
            return Math.pow(1000, i+1) * size / 8;
        }
    }
    // SI 千字节    1000 字节 = 1 千字节
    for (let i = 0; i< metric_prefix_zh.length; i++) {
        if (prefix === `${metric_prefix_zh[i]}字节` || prefix === `SI ${metric_prefix_zh[i]}字节`) {
            return Math.pow(1000, i+1) * size;
        } else if (prefix === `${metric_prefix_zh[i]}比特` || prefix === `SI ${metric_prefix_zh[i]}比特`) {
            return Math.pow(1000, i+1) * size / 8;
        }
    }
    // JEDEC KB    1024 B = 1 KB
    if (prefix === "KB") return Math.pow(1024, 1) * size;
    if (prefix === "Kb") return Math.pow(1024, 1) * size / 8;
    for (let i = 0; i< jedec_prefix.length; i++) {
        if (prefix === `JEDEC ${jedec_prefix[i]}B`) {
            return Math.pow(1024, i+1) * size;
        } else if (prefix === `JEDEC ${jedec_prefix[i]}b`) {
            return Math.pow(1024, i+1) * size / 8;
        }
    }
    return -2;
}

/**
 * 字节转换为包含词缀的文件大小
 * @param {number} byte 字节大小
 * @param {string} prefix 词缀 "B" "kB" "KiB"
 * @param {number} digits 保留小数位数
 * @returns {string} 文件大小 "1 B"
 */
function byteToPrefix(byte=1, prefix="B", digits=null) {
    if (digits !== null && (typeof digits !== "number" || digits < 0)) digits = null;
    /**
     * 自动选择词缀
     * @param {number} byte 字节大小
     * @param {string} prefixType "IEC"," SI", "JEDEC", "SI 中文"]
     * @param {string} basePrefix binary_prefix / metric_prefix / jedec_prefix
     * @param {number} base 1024 或 1000
     */
    function getAutoPrefix(byte, prefixType, basePrefix, base=1024) {
        let i = 0;
        while (byte >= base && i < basePrefix.length) {
            byte /= base;
            i++;
        }
        let b= "B";
        if (prefixType === "IEC") {
            if (i !== 0) {
                b = "iB";
            }
        }else if (prefixType === "SI 中文") {
            b = "字节";
            prefixType = "SI"
        }
        return `${(i === 0) ? '' : prefixType + ' ' + basePrefix[i-1]}${b}`;
    }
    if (prefix === "IEC 自动") {
        prefix = getAutoPrefix(byte, "IEC", binary_prefix, 1024);
    } else if (prefix === "SI 自动") {
        prefix = getAutoPrefix(byte, "SI", metric_prefix, 1000);
    } else if (prefix === "JEDEC 自动") {
        prefix = getAutoPrefix(byte, "JEDEC", jedec_prefix, 1024);
    } else if (prefix === "SI 中文自动") {
        prefix = getAutoPrefix(byte, "SI 中文", metric_prefix_zh, 1024);
    }
    let prefixByte = prefixToByte(prefix);
    if (prefixByte === -1) return "参数错误";
    if (prefixByte === -2) return "未知词缀";
    prefix = prefix.replace(/(SI|IEC|JEDEC) /, "");
    if (digits === null) return "" + (byte / prefixByte) + " " + prefix;
    return (byte / prefixByte).toFixed(digits) + " " + prefix;
}



/**
 * 初始化转换工具
 */
function initConversionTools() {
    // 初始化输入词缀选择框
    const inputPrefixSelectElement = document.getElementById('input-prefix');
    all_prefix.forEach(prefix => {
        const option = document.createElement("option");
        option.value = prefix;
        option.textContent = prefix;
        inputPrefixSelectElement.appendChild(option);
    });
    // inputPrefixSelectElement.selectedIndex = 14;
    // 初始化输出词缀选择框
    const outPrefixSelectElement = document.getElementById('output-prefix');
    auto_prefix.concat(all_prefix).forEach(prefix => {
        const option = document.createElement("option");
        option.value = prefix;
        option.textContent = prefix;
        outPrefixSelectElement.appendChild(option);
    });
    // outPrefixSelectElement.selectedIndex = 0;
    // 监听输入框变化
    const inputSizeInputElement = document.getElementById('input-size');
    const conversionResultSpanElement = document.getElementById('conversion-result');
    /**
     * 更新转换结果
     */
    function updateConversionResult() {
        let inputPrefix = inputPrefixSelectElement.value;
        let outPrefix = outPrefixSelectElement.value;
        let byte = prefixToByte(inputPrefix, Number(inputSizeInputElement.value));
        
        conversionResultSpanElement.textContent = 
            inputPrefix.replace(/^(SI|IEC|JEDEC) /, "") + 
            " = " + byteToPrefix(byte, outPrefix);
    }
    inputSizeInputElement.addEventListener("input", updateConversionResult);
    inputPrefixSelectElement.addEventListener("change", updateConversionResult);
    outPrefixSelectElement.addEventListener("change", updateConversionResult);
    // inputSizeInputElement.value = "1000";
    updateConversionResult();

}

/**
 * 更新显示文件列表
 */
function updateFileList() {
    const uploadPrefixSelectElement = document.getElementById('upload-prefix');
    let uploadResultElement = document.getElementById("upload-result");
    uploadResultElement.innerHTML = "";
    for (let i = fileList.length - 1; i >= 0; i--) {
        const file = fileList[i];
        console.log(file);
        const fileSpanElement = document.createElement("span");
        fileSpanElement.textContent = file.webkitRelativePath ? file.webkitRelativePath : file.name + " - " + byteToPrefix(file.size, uploadPrefixSelectElement.value);
        const fileElement = document.createElement("div");
        fileElement.appendChild(fileSpanElement);
        uploadResultElement.appendChild(fileElement);
    }
}

/**
 * 添加文件到文件列表并显示
 * @param {File[]} files 文件列表
 */
function addFiles(files) {
    for (let i = 0; i < files.length; i++) {
        const file = files[i];
        fileList.push(file);
    }
    updateFileList();
}

/**
 * 初始化拖放选择文件
 */
function initDragDrop() {
    // 使用拖放来选择文件 https://developer.mozilla.org/zh-CN/docs/Web/API/File_API/Using_files_from_web_applications#%E4%BD%BF%E7%94%A8%E6%8B%96%E6%94%BE%E6%9D%A5%E9%80%89%E6%8B%A9%E6%96%87%E4%BB%B6
    function dragenter(e) {
        e.stopPropagation();
        e.preventDefault();
        console.log("dragenter");
    }

    function dragover(e) {
        e.stopPropagation();
        e.preventDefault();
        console.log("dragover");
    }

    function drop(e) {
        e.stopPropagation();
        e.preventDefault();

        const dt = e.dataTransfer;
        const files = dt.files;

        addFiles(files);
        console.log(files);

    }

    const dropbox = document.getElementById("dropbox");
    dropbox.addEventListener("dragenter", dragenter, false);
    dropbox.addEventListener("dragover", dragover, false);
    dropbox.addEventListener("drop", drop, false);
}

/**
 * 初始化上传检测文件大小
 */
function initUpload() {
    // 初始化上传词缀选择框
    const uploadPrefixSelectElement = document.getElementById('upload-prefix');
    auto_prefix.concat(all_prefix).forEach(prefix => {
        const option = document.createElement("option");
        option.value = prefix;
        option.textContent = prefix;
        uploadPrefixSelectElement.appendChild(option);
    });
    // 清空列表按钮
    const clearButtonElement = document.getElementById("clear-button");
    clearButtonElement.addEventListener("click", () => {
        fileList = [];
        updateFileList();
    });
    uploadPrefixSelectElement.addEventListener("change", () => {
        updateFileList();
    });
    // 上传多文件
    const fileInputElement = document.getElementById("file-input");
    fileInputElement.addEventListener("change", () => {
        addFiles(fileInputElement.files);
    });
    // 上传目录
    const directoryInput = document.getElementById("directory-input");
    directoryInput.addEventListener("change", () => {
        addFiles(directoryInput.files);
    });
    // 使用拖放来选择文件
    initDragDrop();

}

document.addEventListener('DOMContentLoaded', function() {
    // 初始化转换工具
    initConversionTools();
    // 初始化上传检测文件大小
    initUpload();
});